src/overlays/actors/ovl_Obj_Switch/z_obj_switch.h

```diff
@@ -5,15 +5,63 @@
 
 struct ObjSwitch;
 
-typedef void (*ObjSwitchActionFunc)(struct ObjSwitch*, GlobalContext*);
+typedef void (*ObjSwitchActionFunc)(struct ObjSwitch* this, GlobalContext* globalCtx);
+typedef void (*ObjSwitchDrawFunc)(struct ObjSwitch* this, GlobalContext* globalCtx);
+typedef void (*ObjSwitchSetupActionFunc)(struct ObjSwitch* this);
+
+typedef enum {
+    /* 0 */ OBJSWITCH_TYPE_FLOOR,
+    /* 1 */ OBJSWITCH_TYPE_FLOOR_RUSTY, // Unused, incomplete implementation
+    /* 2 */ OBJSWITCH_TYPE_EYE,
+    /* 3 */ OBJSWITCH_TYPE_CRYSTAL,
+    /* 4 */ OBJSWITCH_TYPE_CRYSTAL_TARGETABLE,
+    /* 5 */ OBJSWITCH_TYPE_FLOOR_LARGE
+} ObjSwitchType;
+
+typedef enum {
+    /* 0 */ OBJSWITCH_SUBTYPE_ONCE, // Set Switch Flag when activated, stays activated
+    /* 1 */ OBJSWITCH_SUBTYPE_TOGGLE, // Toggle Switch Flag when activated
+    /* 2 */ OBJSWITCH_SUBTYPE_RESET, // Floor type only, set switch flag when pressed, unset when released
+    /* 3 */ OBJSWITCH_SUBTYPE_RESET_INVERTED, // Floor type only, set switch flag when not pressed, unset when pressed
+    /* 4 */ OBJSWITCH_SUBTYPE_SYNC // Crystal type only, syncronizes all crystal fl
+} ObjSwitchSubType;
 
 typedef struct ObjSwitch {
-    /* 0x0000 */ Actor actor;
-    /* 0x0144 */ char unk_144[0x18];
+    /* 0x0000 */ DynaPolyActor dyna;
     /* 0x015C */ ObjSwitchActionFunc actionFunc;
-    /* 0x0160 */ char unk_160[0xF8];
+    /* 0x0160 */ s16 floorSwitchReleaseTimer;
+    /* 0x0162 */ s16 disableCrystalSwitchTimer;
+    /* 0x0164 */ s8 eyeTexIndex;
+    /* 0x0165 */ s8 unk_165;
+    /* 0x0168 */ s32 crystalAnimTimer;
+    /* 0x016C */ Color_RGB8 color;
+    /* 0x016F */ u8 unk_16F;
+    /* 0x0170 */ s8 unk_170;
```

Comment:
```suggestion
    /* 0x0170 */ s8 turnPlayerState;
```

This is a variable that tracks whether or not the player has been rotated after pressing a floor switch